cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ReplicationThrottleHelper.java

```diff
@@ -173,15 +179,28 @@ private void setFollowerThrottledReplicas(String topic, Set<String> replicas) {
   }
 
   private void setThrottledReplicas(String topic, Set<String> replicas, String configKey) {
-    assert (configKey.equals(LEADER_THROTTLED_REPLICAS) || configKey.equals(FOLLOWER_THROTTLED_REPLICAS));
+    if ((!configKey.equals(LEADER_THROTTLED_REPLICAS)) && (!configKey.equals(FOLLOWER_THROTTLED_REPLICAS))) {
+      throw new IllegalArgumentException(
+          String.format("Config key must be either %s or %s", LEADER_THROTTLED_REPLICAS, FOLLOWER_THROTTLED_REPLICAS));
+    }
+
     Properties config = _kafkaZkClient.getEntityConfigs(ConfigType.Topic(), topic);
-    // Merge new throttled replicas with existing configuration values.
-    Set<String> newThrottledReplicas = new TreeSet<>(replicas);
     String oldThrottledReplicas = config.getProperty(configKey);
-    if (oldThrottledReplicas != null) {
-      newThrottledReplicas.addAll(Arrays.asList(oldThrottledReplicas.split(",")));
+    if (oldThrottledReplicas != null && oldThrottledReplicas.trim().equals("*")) {
+      // The existing setup throttles all replica. So, nothing needs to be changed
+      return;
+    }
+
+    if (replicas.size() == 1 && replicas.iterator().next().trim().equals("*")) {
+      config.setProperty(configKey, "*");
+    } else {
```

Comment:
Does this private function ever expected to get `*` as input for a replica in `replicas`? Isn't this dead code?

Reply:
Considering the current state of the codebase, this private function does not take `*` as input. But just looking at this private method itself, we do not know what is going to change on its caller side in the future. Due to my defensive style, I prefer not to make assumptions regarding what callers will or never pass to a method even if the method is private. It might be dead code. But it's safe and handles a very valid case.

Reply:
> But it's safe and handles a very valid case.

The key issue is that this is _not_ a valid case.
This code can generate inconsistent / incorrect behavior -- i.e. replication throttler shall never set this config to `*`.

This is not making the code defensive, on the contrary, it makes it more vulnerable to errors.

* We also seem to be missing tests for the added fix -- can we add unit tests please?

Reply:
I'm not following exactly, but given the bug happened where a `*` and a `kv-list` were joined (thus this PR), it well might be that this method gets called with a `*` member, isn't it?

https://kafka.apache.org/documentation/#leader.replication.throttled.replicas and the Kafka validation code certainly allows for `*`. If CC should not allow for a `*` in this method, what is the codepath that still smuggled it? Or otherwise how could the bug happen?

Agree the unit-tests would be welcome nevertheless.